internal/cli/cli.go

```diff
@@ -38,83 +42,180 @@ type CommandDefinition struct {
 }
 
 // CommandFactory is a command factory
-type CommandFactory struct {
-	Profile *Profile
-	Config  *Config
+type CommandFactory interface {
+	Build(func() CommandDefinition) *cobra.Command
+	Close()
+	Init()
+	Run(cmd *cobra.Command)
+	SetGlobalFlags(fs *flag.FlagSet)
+}
+
+type commandFactory struct {
+	in      *os.File
+	out     *os.File
+	err     *os.File
+	config  *Config
+	profile *Profile
+	ui      terminal.UI
+	logger  *log.Logger
 }
 
 // Config is the global CLI config
 type Config struct {
-	Command CommandConfig
-	UI      terminal.UIConfig
+	CommandConfig
+	terminal.UIConfig
 }
 
 // CommandConfig holds the global config for a CLI command
 type CommandConfig struct {
 	RealmBaseURL string
 }
 
+// NewCommandFactory creates a new command factory
+func NewCommandFactory() CommandFactory {
+	logger := log.New(os.Stderr, "UTC ERROR ", log.Ltime|log.Lmsgprefix)
+
+	config := new(Config)
+
+	profile, profileErr := NewDefaultProfile()
+	if profileErr != nil {
+		logger.Fatal(profileErr)
+	}
+
+	return &commandFactory{
+		config:  config,
+		profile: profile,
+		logger:  logger,
+	}
+}
+
+func (factory *commandFactory) Init() {
```

Comment:
could we return an error here instead of being a void func?

Reply:
in theory it could, but it would still need to be wrapped by a void func to be passed into `cobra.OnInitialize`, a hook used to run stuff before any of the `*RunE` command phases (but ran after flags have been parsed).